Don't do detailed analytics on hidden tables

thearthur opened the pull request:
Select only tables that are not marked hidden when choosing tables to analyze.

Resolves #2127

thearthur committed:
Don't do detailed analytics on hidden tables

When symcing a database filter out any tables that have been marked
"hidden" so we don't wast time looking at them.

salsakran commented:
this takes care of 1+3 of list of requirements in @2127
```
Hidden tables shouldn't be sync'd
When a table is unhidden, it should be sync'd
New tables still get sync'd and displayed
```
It would be nice to sync/analyze tables that get unhidden.

thearthur commented:
I'll add the sync on change, and I'd also like to make the select then filter pattern just a select. So a little more work on this one.

thearthur committed:
More efficiently only analyze visible tables

filter in the DB call rather than fitering after fetching

camsaul commented on src/metabase/sync_database/analyze.clj:
```diff
@@ -246,7 +246,13 @@
+                                         [:= :active true]
+                                         [:or
+                                          [:= :visibility_type nil]
+                                          [:not= :visibility_type "hidasdfden"]]]})
```
```clojure
[:not= :visibility_type "hidasdfden"]
```

What?

thearthur replied:
That was an accident with git, has been fixed

camsaul reviewed (changes requested):
This looks like a typo

```clojure
[:not= :visibility_type "hidasdfden"]
```

thearthur committed:
resync tables when visibility changes

thearthur commented:
@camsaul the type has been "un pushed"

camsaul commented on src/metabase/api/table.clj:
```diff
@@ -44,15 +48,21 @@
-                   :description             description))
-  (api/check-500 (db/update! Table id, :visibility_type visibility_type))
-  (Table id))
+  (let [original-visibility_type (:visibility_type (Table :id id))]
```
Since the API and the DB use `snake_case` what to name something like this is a tricky question. But since it's not being used by the API or the DB I would just keep it `lisp-case` and name it `original-visibility-type`. Mixing the two just looks weird :D

camsaul commented on src/metabase/api/table.clj:
```diff
@@ -44,15 +48,21 @@
+                     :entity_type             entity_type
+                     :description             description))
+    (api/check-500 (db/update! Table id, :visibility_type visibility_type))
+    (let [updated-table (Table :id id)
```
You can just do `(Table id)` to fetch the Table whose `:id` is `id`.

camsaul commented:
So it looks like it calls `(sync-database/sync-table! updated-table)` when you change the visibility type, but I don't see any new logic in `sync-table!` to skip the table based on its visibility type? Plus I think we only want to call that if the table goes from an "unsynced" visibility type to a "synced" one. e.g. there's no need to resync if we change it from a "synced" visibility type to another synced one.

Open question: Which types get synced? There's effectively four options:

```clojure
(def ^:const visibility-types
  "Valid values for `Table.visibility_type` (field may also be `nil`)."
  #{:hidden :technical :cruft})
```

I think we might want to skip `:cruft` tables as well. But if we sync either `nil` or `:technical` there's no need to sync when changing between the two

thearthur commented:
good question, it's also asked at the bottom of the enclosing issue. 
As there is more detail here, let's address it on this end:

sure! let's call lumping hidden and cruft into non-sync-things
and nil (visible) and :technical into things that get synced.

Sound good?

camsaul commented:
I think syncing `nil` (visible) and`:technical` tables but not `:hidden` and `:cruft` ones sounds right. @salsakran does that make sense?

thearthur commented:
On second Though :technical :cruft and :hidden are all states under the "non queryable" option in the UI. It makes sense to not analyze them if they are not something a person can query

camsaul commented:
Oh I forgot you couldn't see `:technical` tables. Basically the way things are now `:visibilty_type` would better be called `:reason_for_hiding`

thearthur commented:
so only things that have `nil` reason to hide should then be analyzed

camsaul commented:
yeah

salsakran commented:
one of the reasons to separate out hidden tables into different subtypes is to let us auto-detect certain classes of them (eg, rails framework tables) and set their class to a hidden subtype and not just "hidden", so the user knows why.

thearthur committed:
Don't analyze :technical or :cruft tables either

only sync tables in the queryable state.
don't re-sync on transitions between different hidden states.

thearthur commented:
OK now only nil is considered visible and it only syncs on transition between visible and non-visible

salsakran commented:
can we get a test up in this?

thearthur committed:
slight code cleanup

thearthur committed:
add tests for re-sync on visibility change

salsakran commented:
@thearthur not sure when the best time to do this, but can we also have a test that verifies that we only sync visible tables? The test here verifies that the API calls the sync function when a table is unhidden, but there's nothing verifying that hidden tables are skipped.

I'd suggest creating a two tables where one is hidden, and verifying that the visible on got sync'd and the other didn't.

camsaul commented on src/metabase/api/table.clj:
```diff
@@ -37,22 +41,40 @@
       (hydrate :db :pk_field)))
 
+(defn- visible-state?
+  "only the nil state is considered visible."
```
Since this function ends in a `?` it sort of implies it returns a truthy or falsy value. Maybe a name like `visibility-type->visibility-state` would be more appropriate or even just making this return `true` instead of `:show` and `false` instead of `hide` would make it a lot simpler

camsaul commented:
@arthur one way you could do a test like @salsakran suggests is to make a fake DB copying the details from sample dataset and then sync it (which will cause it to get copies of the tables and fields). Then change one of the tables to `:hidden` and change its name to a different table (e.g. change the `venues` table to `checkins`) and sync again. Then see if the `:hidden` table's row count is updated

thearthur commented:
A test very similar to the one added in the last check may be easier by testing if the sync function runs. so
* make it visible
* do a sync on that db
* make it all variations on hidden
* do a sync between each
* make it visible
* do a sync
* check that it has been synced twice

salsakran commented:
I don't think testing if the sync functions runs is enough. We should verify the specific table has been sync'd.

salsakran commented:
In general, our philosophy on testing is to answer the question "did the features in this PR get broken by any subsequent change" not "do the changes in this PR work".
Verification that it worked is the lowest bar, the real safety net is "can this bug re-appear" or the feature equivalent.

camsaul commented:
another option is

*  Create a temp DB copying the details of the test DB
*  Create a temp table belonging to the temp DB with one of the test DB table names, like `venues`. Make it `:hidden`
*  Sync the temp DB
*  The `venues` table shouldn't have any Fields (didn't get synced) but everything else should

thearthur commented:
It looks like there are two phases, Sync and Analyze. 
the first phase is creating the Field entries in the DB so checking for these is not a useful indication of Analysis happening or not.

there are a couple options for this:
1. test it less efectivly
2. improve this feature so it also covers not Syncing tables that are hidden in addition to not analyzing them.
3. test it some other way and let it continue to sync the tables even through it does not do further analysis on them.

#2 produces the best experience for people using metabase, so I vote for that. It makes things faster and doesn't require explaining things to people. I don't know if not syncing them will have side effects elsewhere @camsaul  is that something i should worry about?

salsakran commented:
how about:
1. start with a hidden table in the temp db
2. flip it to being visible
3. see if it has a row count now?
